clippy_lints/src/map_nil_fn.rs

```diff
@@ -0,0 +1,187 @@
+use rustc::hir;
+use rustc::lint::*;
+use rustc::ty;
+use syntax::codemap::Span;
+use utils::{in_macro, iter_input_pats, match_type, method_chain_args, snippet, span_lint_and_then};
+use utils::paths;
+
+#[derive(Clone)]
+pub struct Pass;
+
+/// **What it does:** Checks for usage of `Option.map(f)` where f is a nil
+/// function or closure
+///
+/// **Why is this bad?** Readability, this can be written more clearly with
+/// an if statement
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+/// ```rust
+/// let x : Option<&str> = do_stuff();
+/// x.map(log_err_msg);
+/// x.map(|msg| log_err_msg(format_msg(msg)))
+/// ```
+/// The correct use would be:
+/// ```rust
+/// let x : Option<&str> = do_stuff();
+/// if let Some(msg) = x {
+///     log_err_msg(msg)
+/// }
+/// if let Some(msg) = x {
+///     log_err_msg(format_msg(msg))
+/// }
+/// ```
+declare_lint! {
+    pub OPTION_MAP_NIL_FN,
+    Allow,
```

Comment:
Can this be `Warn`? I'm not sure, since `.map()` is also a rather common construct even with unit functs, and I'm not sure if `if let` really is unambiguously better. I personally use `if let` as much as possible, though it took me some time to switch over to that (and you'll see a lot of pre-2014 code using it).

thoughts @llogiq @mcarton